flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java

```diff
@@ -877,4 +899,610 @@ private String getSensorPositionInfo(Asset asset) {
       }
       return positionInfo.toString();
    }
+
+   private Map<Asset, List<ValueUnit>> getPduValueUnitsMap(Asset pduAsset, long starttime, int duration){
+      List<ValueUnit> pduValueunits = getValueUnitsByID(pduAsset.getId(), starttime, duration);
+      Map<Asset, List<ValueUnit>> assetAndValueUnitsMap = new HashMap<Asset, List<ValueUnit>>();
+      if(!pduValueunits.isEmpty()) {
+         //Filter pdu metric by metric name
+         List<String> metricNames = new ArrayList<String>();
+         metricNames.add(MetricName.PDU_TOTAL_POWER);
+         metricNames.add(MetricName.PDU_TOTAL_CURRENT);
+         metricNames.add(MetricName.PDU_APPARENT_POWER);
+         metricNames.add(MetricName.PDU_ACTIVE_POWER);
+         metricNames.add(MetricName.PDU_CURRENT);
+         metricNames.add(MetricName.PDU_VOLTAGE);
+         metricNames.add(MetricName.PDU_FREE_CAPACITY);
+         metricNames.add(MetricName.PDU_POWER_LOAD);
+         metricNames.add(MetricName.PDU_CURRENT_LOAD);
+         filterValueUnitsByMetricNames(pduValueunits, metricNames);
+         assetAndValueUnitsMap.put(pduAsset, pduValueunits);
+      }
+      //sensor valueUnits, such as temperature or humidity
+      Map<String, String> formulars = pduAsset.getMetricsformulars();
+      String sensorFormulasInfo = formulars.get(FlowgateConstant.SENSOR);
+      Map<String, Map<String, String>> sensorFormulasMap = null;
+      if(sensorFormulasInfo != null) {
+         sensorFormulasMap =
+               pduAsset.metricsFormulaToMap(sensorFormulasInfo, new TypeReference<Map<String, Map<String, String>>>() {});
+         if(sensorFormulasMap.isEmpty()) {
+            return assetAndValueUnitsMap;
+         }
+         Set<String> assetIds = getSensorAssetIDFromFormula(sensorFormulasMap);
+         for(String sensorId : assetIds) {
+            List<ValueUnit> valueUnits = getValueUnitsByID(sensorId, starttime, duration);
+            if(!valueUnits.isEmpty()) {
+               Asset sensor = getAssetById(sensorId);
+               assetAndValueUnitsMap.put(sensor, valueUnits);
```

Comment:
Do we need to remove the uneeded metrics from the sensor asset?  for example if one sensor has multiple metrics but PDU only need the Temperature and humidity?